File: my-folder/0074-search-a-2d-matrix/solution.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        rows,cols=len(matrix),len(matrix[0])
        r,c=0,cols-1
        while r<rows and c>=0:
            if matrix[r][c]>target:
                c-=1
            elif matrix[r][c]<target:
                r+=1
            else:
                return True
        return False



        

File: my-folder/0074-search-a-2d-matrix/solution.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int top = 0;
        int bot = matrix.length - 1;
        int cols = matrix[0].length;
        while(top <= bot){
            int row = (top + bot) / 2;
            if(matrix[row][cols - 1] < target){
                top = row + 1;
            }
            else if(matrix[row][0] > target){
                bot = row - 1;
            }
            else{
                break;
            }
        }

        if (!(top <= bot)) {
        return false;
        } 

         int row = (top + bot) / 2;
        int l = 0, r = cols - 1;
        while (l <= r) {
            int m = (l + r) / 2;
            if (target > matrix[row][m]) {
                l = m + 1;
            } else if (target < matrix[row][m]) {
                r = m - 1;
            } else {
                return true;
            }
        }
        return false;
        
    }
}
